llvm/lib/Target/DirectX/DXILRootSignature.h

```diff
@@ -0,0 +1,66 @@
+//===- DXILRootSignature.h - DXIL Root Signature helper objects -----------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file This file contains helper objects and APIs for working with DXIL
+///       Root Signatures.
+///
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Analysis/DXILMetadataAnalysis.h"
+#include "llvm/IR/DiagnosticInfo.h"
+#include "llvm/IR/Metadata.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+#include <optional>
+
+namespace llvm {
+namespace dxil {
+
+enum class RootSignatureElementKind { None = 0, RootFlags = 1 };
+
+struct ModuleRootSignature {
+  uint32_t Flags = 0;
+  static std::optional<ModuleRootSignature> analyzeModule(Module &M,
+                                                          const Function *F);
```

Comment:
This doesn't really need to be public API like this - better to just have a `static` analyzeModule function in DXILRootSignature.cpp rather than a static member of the ModuleRootSignature struct.